src/audio/google/google_rtc_audio_processing.c

```diff
@@ -83,30 +110,131 @@ void GoogleRtcFree(void *ptr)
 	return rfree(ptr);
 }
 
-#if CONFIG_IPC_MAJOR_4
-static void google_rtc_audio_processing_params(struct processing_module *mod)
+static ALWAYS_INLINE float s16_to_float(const char *ptr)
 {
-	struct google_rtc_audio_processing_comp_data *cd = module_get_private_data(mod);
-	struct sof_ipc_stream_params *params = mod->stream_params;
-	struct comp_buffer *sinkb, *sourceb;
-	struct list_item *source_list;
-	struct comp_dev *dev = mod->dev;
+	float scale = -(float)SHRT_MIN;
+	float x = *(int16_t *)ptr;
 
-	ipc4_base_module_cfg_to_stream_params(&mod->priv.cfg.base_cfg, params);
-	component_set_nearest_period_frames(dev, params->rate);
+	return (1.0f / scale) * x;
+}
+
+static ALWAYS_INLINE void float_to_s16(float x, char *dst)
+{
+	float scale = -(float)SHRT_MIN;
+	float min = -1.0f;
+	float max = 1.0f - 1.0f / scale;
+	int16_t i = (int16_t)(scale * (x < min ? min : (x > max ? max : x)));
+
+	*(int16_t *)dst = i;
+}
+
+static ALWAYS_INLINE float s32_to_float(const char *ptr)
+{
+	float scale = -(float)INT_MIN;
+	float x = *(int32_t *)ptr;
 
-	list_for_item(source_list, &dev->bsource_list) {
-		sourceb = container_of(source_list, struct comp_buffer, sink_list);
-		if (IPC4_SINK_QUEUE_ID(buf_get_id(sourceb)) == SOF_AEC_FEEDBACK_QUEUE_ID)
-			ipc4_update_buffer_format(sourceb, &cd->config.reference_fmt);
-		else
-			ipc4_update_buffer_format(sourceb, &mod->priv.cfg.base_cfg.audio_fmt);
+	return (1.0f / scale) * x;
```

Comment:
`x / scale`?

Reply:
Style.  Writing the former makes it clear that we are generating a runtime multiplication by a compile-time constant.

Reply:
Actually not style.  Floating point won't refactor in the optimizer like that because the precision behavior is different.  So writing `x / scale` would produce the runtime division that we're desperately trying to avoid.